### ISSUE TITLE
Running Single Image on LoFTR

### ISSUE THREAD

**Issue by atharvjairath** [zju3dv/LoFTR#218]
Hey,
Great work with LoFTR!
```python
def forward(self, data):
        """ 
        Update:
            data (dict): {
                'image0': (torch.Tensor): (N, 1, H, W)
                'image1': (torch.Tensor): (N, 1, H, W)
                'mask0'(optional) : (torch.Tensor): (N, H, W) '0' indicates a padded position
                'mask1'(optional) : (torch.Tensor): (N, H, W)
            }
        """
        # 1. Local Feature CNN
        data.update({
            'bs': data['image0'].size(0),
            'hw0_i': data['image0'].shape[2:], 'hw1_i': data['image1'].shape[2:]
        })

        if data['hw0_i'] == data['hw1_i']:  # faster & better BN convergence
            feats_c, feats_f = self.backbone(torch.cat([data['image0'], data['image1']], dim=0))
            (feat_c0, feat_c1), (feat_f0, feat_f1) = feats_c.split(data['bs']), feats_f.split(data['bs'])
        else:  # handle different input shapes
            (feat_c0, feat_f0), (feat_c1, feat_f1) = self.backbone(data['image0']), self.backbone(data['image1'])

        data.update({
            'hw0_c': feat_c0.shape[2:], 'hw1_c': feat_c1.shape[2:],
            'hw0_f': feat_f0.shape[2:], 'hw1_f': feat_f1.shape[2:]
        })

        # 2. coarse-level loftr module
        # add featmap with positional encoding, then flatten it to sequence [N, HW, C]
        feat_c0 = rearrange(self.pos_encoding(feat_c0), 'n c h w -> n (h w) c')
        feat_c1 = rearrange(self.pos_encoding(feat_c1), 'n c h w -> n (h w) c')

        mask_c0 = mask_c1 = None  # mask is useful in training
        if 'mask0' in data:
            mask_c0, mask_c1 = data['mask0'].flatten(-2), data['mask1'].flatten(-2)
        feat_c0, feat_c1 = self.loftr_coarse(feat_c0, feat_c1, mask_c0, mask_c1)

        # 3. match coarse-level
        self.coarse_matching(feat_c0, feat_c1, data, mask_c0=mask_c0, mask_c1=mask_c1)

        # 4. fine-level refinement
        feat_f0_unfold, feat_f1_unfold = self.fine_preprocess(feat_f0, feat_f1, feat_c0, feat_c1, data)
        if feat_f0_unfold.size(0) != 0:  # at least one coarse level predicted
            feat_f0_unfold, feat_f1_unfold = self.loftr_fine(feat_f0_unfold, feat_f1_unfold)

        # 5. match fine-level
        self.fine_matching(feat_f0_unfold, feat_f1_unfold, data)
```
By default we require pair of images for inference, I want to precompute some part of the inference for an anchor image in the pair, so that when we run inference on query and anchor image as pair, it only need to process some inference of query (do improve latency).

Therefore I wanted to know in the above forward function, till where this two input (query and anchor ) are independent of the output. For what I know:
1. **Local Feature CNN** is Independent of each other
2. **coarse-level loftr module** seems like it is independent of each other
3.  For **match coarse-level** and  **fine-level refinement** they are dependent, i.e we need both the images together to get output.

please correct me if I am wrong, and it would be great if you can comment on this approach and suggest some changes.

Thank you!

**Comment by zehongs** [comment]
Hi, sorry for the late reply.
I'm not sure if I understood your question correctly. The two feature is dependent on each other after the **coarse-level loftr module**.  And since there is a self-attention layer in the loftr-module, predefining an interest region in one image and masking all other parts may harm the performance. So, I'm quite interested in your experiments. Let me know if you have any other questions!
As for latency, I happen to have some new tests that I can share: LoFTR-DS runs at 55ms with a peak memory cost of 1079MB for a 640 × 480 image pair on an RTX 3090. Specifically, the local feature CNN takes 28ms, the coarse-level LoFTR 12ms, the DS layer 7ms, and the coarse-to-fine module 8ms.
